### PR TITLE
Bundled Dependencies

### PR DESCRIPTION
**Pull request by rgrinberg**
This PR introduces a bundled dependency type. I haven't yet written
a concrete syntax because I'm not sure which is better:

Marking the individual deps

```
(:bundle foo)
```

Or marking groups of deps:

```
(:bundle foo (bar >= 0.2))
```

@diml any preference?

cc @avsm

### CODE REVIEW

**Comment by diml** [comment]
What about `(foo :bundled)`? That would compose nicely with the other `:`-variables: `(foo :bundled :test)`.

**Comment by avsm** [comment]
+1 to @diml's suggestion of `(foo :bundled)`

**Comment by rgrinberg** [comment]
Sounds fine but we have to be a bit careful here. This variable isn't like other opam variables so it cannot be evaluated conditionally and we'd have a bit of a problem if opam introduces such a variable. I suggest that we only allow in the first position. E.g:

```
(foo :bundle <constraints>)
```

Also, let's not use the past tense. It's really the user that is asking dune to bundle this package. `:bundle` also reads more natural to me.

**Comment by rgrinberg** [comment]
@diml ready now

File: test/blackbox-tests/test-cases/bundle/only-packages/run.t [hunk 1]
```diff
@@ -0,0 +1,26 @@
+  $ cat >dune-project <<EOF
```
Comment (diml) [inline test/blackbox-tests/test-cases/bundle/only-packages/run.t]: Could you add a comment to explain what this test is about?

**Comment by diml** [comment]
BTW, it seems to me that the `@install` alias (and so `-p`) will still build the `.install` files of bundled packages, and so will build these packages entirely. Do you agree? That seems a bit unfortunate. We should build only what we need from bundled packages.

Another thought: what happens for bundled dependencies of direct bundled dependencies?

i.e.:

```scheme
(package
 (name a))

(package
 (name b)
 (depends
  (a :bundle)))

(package
 (name c)
 (depends
  (b :bundle)))
```

**Comment by rgrinberg** [comment]
> BTW, it seems to me that the @install alias (and so -p) will still build the .install files of bundled packages, and so will build these packages entirely. Do you agree? That seems a bit unfortunate. We should build only what we need from bundled packages.
Yeah, that’s a bit of a shame but I’m not sure how to fix this with the current scheme. Perhaps if we strict_package_dirs, we could make -p build `@pkg/install`.
> Another thought: what happens for bundled dependencies of direct bundled dependencies?
At first I thought that making bundling transitive would be the right thing here. But in fact, that does introduce some limitations.

For example, if b bundles a only to use it as a binary to generate some code, then in fact, c should not have a bundled. Such a build however would require per package scopes which we don’t currently have.

On the other hand, if b defines a library a, then the bundling should definitely be transitive.

How about we just restrict bundling dependencies that bundle other dependencies for now?

**Comment by diml** [comment]
> Yeah, that’s a bit of a shame but I’m not sure how to fix this with the current scheme. Perhaps if we strict_package_dirs, we could make -p build `@pkg/install`.

What about not adding the `.install` file of packages that were pulled through bundled dependencies to the `@install` alias?

> How about we just restrict bundling dependencies that bundle other dependencies for now?

So the example I wrote would produce an error because `c` bundles `b` and `b` has bundled dependencies? That seems good to me

**Comment by diml** [comment]
> So the example I wrote would produce an error because c bundles b and b has bundled dependencies? That seems good to me

And at least, it's a future proof first approximation

**Comment by rgrinberg** [comment]
I just discussed this feature with Arseniy & Andrey over lunch. Here’s a summary of the feedback:


• bundle is a confusing name and “private” was suggested instead. I tend to agree with this.
• It was pointed out that the this implementation of bundling does not invalidate some incorrect builds. For example, consider public library a in package A that depends on public library b in B, where A bundles B. This is incorrect.
• Given the issue above and transitive bundling, it was proposed that per-package scopes should fix both issues. For the example above, libraries defined in B should use a scope where all libraries from A are private.

**Comment by rgrinberg** [comment]
@diml sorry I force pushed over your commits. Could you push them again?

**Comment by diml** [comment]
Sure, but maybe they make less sense with the new choices. I'm not sure I understand what are per-package scopes BTW, let's discuss it at some point

**Comment by rgrinberg** [comment]
We've had an offline discussion about the issues above and it was decided that while the name is confusing public/private aren't quite the same thing. public/private are used to mark the artifacts (nodes) themselves, while this marks a dependency (edge) between them.

Regarding not allowing libraries to leak out bundled deps, we'll add this check later.

**Pull request closed without merging**